docs/source/architecture.rst

```diff
@@ -1,69 +1,80 @@
 
 For more details on how to configure the templater see :ref:`templateconfig`.
 
 
 Stage 2, the lexer
 ------------------
 
-The lexer takes raw text and splits it into tokens. Nothing is *removed* but
-whitespace and code are separated. In principle all identifiers should be
-separated at this stage, but they should not be imparted any meaning at this
-stage. Any files which cannot be lexed, should raise a *SQLLexError*.
+The lexer takes valid SQL and separates it into segments of whitespace and
+code. No meaning is imparted, that is the job of the parser.
 
-While the Lexer is passed a series of raw segments, it will return a single
-segment, usually a :code:`FileSegment`, which will be used to initiate parsing.
 
 Stage 3, the parser
 -------------------
 
-We recursively parse each of the elements, using their in built grammars.
-
-#. Initially we start with a segment, which contains only raw segments. This
-   is normally a :code:`FileSegment`, but it could in theory be any kind of
-   segment, and as we recurse it will be lower and lower sublevels.
-
-#. We then *parse* this element by calling :code:`.parse()`.
-
-   #. This first calls uses the :code:`parse_grammar` if that is present, on
-      which we call :code:`.match()`. The *match* method of this grammar will
-      return a potentially refined structure of the segments within this
-      segment in greater detail than what was initially there. In the example
-      of a :code:`FileSegment`, it first divides up the query into statements
-      and then finishes.
-
-   #. The :code:`.match()` method of any grammar is naturally recursive, and
-      is comprised of *segments* and *grammars*. The match step is still not
-      inert however and child elements of a match grammar may return more
-      specific versions of segments during this phase even though we are not
-      calling :code:`.parse()`
-
-      * *Segments* must implement a :code:`match_grammar` to be used in this
-        way. When :code:`.match()` is called on a segment, this is the grammar
-        which is used to decide whether there is a match.
-
-      * *Grammars* are objects which combine *segments* or other *grammars*
-        together in a pre-defined way. For example the :code:`OneOf` grammar
-        will match if any one of its child elements match.
+The parser is arguably the most complicated element of SQLFluff, and is
+relied on by all the other elements of the tool to do most of the heavy
+lifting.
+
+#. The lexed segments are parsed using the specified dialect's grammars. In
+   SQLFluff, grammars describe the shape of SQL statements (or their
+   components). The parser attempts to apply each potential grammar to the
+   lexed segments until all the segments have been matched.
+
+#. In SQLFluff, segments form a tree-like structure. The top-level segment is
+   a :code:`FileSegment`, which contains zero or more
+   :code:`StatementSegment`\ s, and so on. Before the segments have been parsed
```

Comment:
Is the `\ s` deliberate? (I'm not a markdown expert, so I wasn't sure if this was correct or not.)

Reply:
Yep, from reading it seems that the escaped whitespace tells the doc compiler that the code segment has ended. I'm also not a markdown expert, but it works like this!